src/compiler/core.ts

```diff
@@ -2882,3 +2882,56 @@ export function isNodeLikeSystem(): boolean {
         && !(process as any).browser
         && typeof module === "object";
 }
+
+
+/** @internal */
+export interface StackSet<T extends {}> {
+    /**
+     * Returns true if the value has already been pushed.
+     */
+    has(value: T): boolean;
+    /**
+     * Pushes to the stack. The value pushed must not have been pushed before.

```

Comment:
Risky! Name the method something that makes this clearer if you're really intent on doing this instead of just introducing a local `Set`.

Reply:
Can you come up with a better name? I tried over and over and did not come up with anything helpful.

I could just not have this and go back to loops.

Reply:
Oops, sorry, I thought this comment was on the `pop` line.

I could add the debug checks back and see if they make it slower.

Reply:
Debug assert appears to be fine.